1. У нас є дані про час перебування на двох різних сторінках продукту. **Проведемо тест Манна-Вітні і визначимо, чи є різниця в медіанах часу перебування на сторінках.**

 Початкові дані:
 ```
 page_a_time = [54.97, 48.62, 56.48, 65.23, 47.66, 47.66, 65.79, 57.67, 45.31, 55.43,
                  45.37, 45.34, 52.42, 30.87, 32.75, 44.38, 39.87, 53.14, 40.92, 35.88]
 page_b_time = [72.59, 52.29, 55.81, 37.9,  48.47, 56.33, 41.19, 59.51, 47.79, 51.5,  
                  47.78, 77.23, 54.84, 42.31, 64.87, 40.35, 57.51, 31.48, 39.06, 57.36]
 ```

In [3]:
from scipy.stats import mannwhitneyu


# Data on time spent on pages (in seconds)
time_on_page_A = [54.97, 48.62, 56.48, 65.23, 47.66, 47.66, 65.79, 57.67, 45.31, 55.43,
                  45.37, 45.34, 52.42, 30.87, 32.75, 44.38, 39.87, 53.14, 40.92, 35.88]
time_on_page_B = [72.59, 52.29, 55.81, 37.9,  48.47, 56.33, 41.19, 59.51, 47.79, 51.5,
               47.78, 77.23, 54.84, 42.31, 64.87, 40.35, 57.51, 31.48, 39.06, 57.36]


# Perform the Mann-Whitney U test
stat, p_value = mannwhitneyu(time_on_page_A, time_on_page_B)

print(f"U-statistic: {stat}")
print(f"P-value: {p_value}")


# Interpretation
if p_value < 0.05:
  print("There is a significant difference in time spent on pages A and B.")
else:
  print("There is no significant difference in time spent on pages A and B.")

U-statistic: 166.0
P-value: 0.3648191601135933
There is no significant difference in time spent on pages A and B.


**Висновок**:\
Для порівняння часу, проведеного користувачами на сторінках A та B, було застосовано непараметричний критерій Манна — Вітні. Це дозволяє оцінити, чи існує статистично значуща різниця між двома незалежними вибірками без припущення про нормальний розподіл даних.

Отримане значення статистики U дорівнює 166.0, а p-значення становить 0.3648. Оскільки p-значення значно перевищує загальноприйнятий рівень значущості α = 0.05, немає підстав відхиляти нульову гіпотезу про однаковість розподілів.

**Отже, статистично значущої різниці у часі перебування користувачів на сторінках A та B не виявлено. Результати експерименту не дають підстав вважати, що одна зі сторінок утримує увагу користувачів довше за іншу.**

2. Дано три сторінки продукту, і необхідно перевірити, чи є значна різниця в часі, який користувачі проводять на цих сторінках. Проведемо тест Крускала-Уолліса і зробимо висновки. На якій сторінці користувачі перебували найдовше?

 Початкові дані:
 ```
 page_a_seconds = [54.97, 48.24, 59.98, 67.24, 46.91, 46.98, 66.48, 54.35, 42.78, 53.37,
                     43.55, 66.30, 62.28, 42.91, 53.96, 43.61, 59.75, 50.14, 44.56, 48.59]
 page_b_seconds = [59.71, 60.18, 43.29, 58.08, 50.91, 64.05, 54.33, 63.28, 68.80, 50.65,
                     64.79, 72.79, 49.77, 65.23, 48.73, 45.54, 54.47, 57.28, 61.83, 50.38]
 page_c_seconds = [73.41, 44.78, 67.57, 62.83, 55.82, 80.32, 43.27, 71.05, 65.68, 43.36,
                     75.98, 46.57, 69.03, 55.65, 85.73, 58.39, 69.28, 56.21, 60.14, 53.19]
 ```

In [10]:
from scipy.stats import kruskal
from statistics import mean


# Data on time spent on three pages (in seconds)
page_A_seconds = [54.97, 48.24, 59.98, 67.24, 46.91, 46.98, 66.48, 54.35, 42.78, 53.37,
                  43.55, 66.30, 62.28, 42.91, 53.96, 43.61, 59.75, 50.14, 44.56, 48.59]
page_B_seconds = [59.71, 60.18, 43.29, 58.08, 50.91, 64.05, 54.33, 63.28, 68.80, 50.65,
                  64.79, 72.79, 49.77, 65.23, 48.73, 45.54, 54.47, 57.28, 61.83, 50.38]
page_C_seconds = [73.41, 44.78, 67.57, 62.83, 55.82, 80.32, 43.27, 71.05, 65.68, 43.36,
                  75.98, 46.57, 69.03, 55.65, 85.73, 58.39, 69.28, 56.21, 60.14, 53.19]

# Perform the Kruskal-Wallis test
stat, p_value = kruskal(page_A_seconds, page_B_seconds, page_C_seconds)

print(f"Kruskal-Wallis statistic: {stat}")
print(f"P-value: {p_value}")

# Interpretation
if p_value < 0.05:
    print("There is a significant difference in time spent on different pages.")
else:
    print("There is no significant difference in time spent on the pages.")


# Mean (середнє арифметичен) for each group
mean_A = mean(page_A_seconds)
mean_B = mean(page_B_seconds)
mean_C = mean(page_C_seconds)

print(f"Average time on page A: {mean_A:.2f} секунд")
print(f"Average time on page B: {mean_B:.2f} секунд")
print(f"Average time on page C: {mean_C:.2f} секунд")

Kruskal-Wallis statistic: 6.941639344262285
P-value: 0.031091535348550105
There is a significant difference in time spent on different pages.
Average time on page A: 52.85 секунд
Average time on page B: 57.20 секунд
Average time on page C: 61.91 секунд


**Висновок:**\
В цьому випадку тест показав, що є статистично значуща різниця в часі перебування на сторінках. Найбільше часу користувачі провели на третьому варіанті сторінки С.

3. Необхідно порівняти ефективність двох рекламних кампаній для однієї і тієї ж групи користувачів. Виконаємо тест Вілкоксона і визначимо, чи є значна різниця в конверсії до і після зміни рекламної стратегії.

 Початкові дані:
 ```
 conversion_before = [2.75, 2.49, 2.97, 3.65, 2.41, 2.41, 3.10, 2.89, 2.15, 2.68,
                        2.18, 3.63, 3.39, 2.15, 2.69, 2.16, 3.02, 2.53, 2.26, 2.47]
 conversion_after = [3.13, 3.07, 3.55, 2.92, 3.28, 3.65, 2.69, 2.85, 3.44, 3.16,
                       3.62, 3.97, 2.45, 3.53, 2.65, 2.81, 2.90, 3.19, 3.08, 2.69]
 ```

In [14]:
from scipy.stats import wilcoxon


# Data on conversion rates
conversion_before = [2.75, 2.49, 2.97, 3.65, 2.41, 2.41, 3.10, 2.89, 2.15, 2.68,
                     2.18, 3.63, 3.39, 2.15, 2.69, 2.16, 3.02, 2.53, 2.26, 2.47]
conversion_after = [3.13, 3.07, 3.55, 2.92, 3.28, 3.65, 2.69, 2.85, 3.44, 3.16,
                    3.62, 3.97, 2.45, 3.53, 2.65, 2.81, 2.90, 3.19, 3.08, 2.69]


# Perform the Wilcoxon test
stat, p_value = wilcoxon(conversion_before, conversion_after)

print(f"Wilcoxon statistic: {stat}")
print(f"P-value: {p_value}")

# Interpretation
if p_value < 0.05:
    print("Conversion has significantly improved.")
else:
    print("There is no significant difference in conversion before and after.")


Wilcoxon statistic: 42.0
P-value: 0.017181396484375
Conversion has significantly improved.


**Висновок:**

Для перевірки гіпотези щодо ефективності зміни рекламної кампанії було застосовано парний непараметричний критерій Вілкоксона. Цей тест використовується для порівняння двох залежних вибірок, зокрема — конверсій до та після внесення змін у реклмну кампанію.

Результати тесту:

Статистика критерію Вілкоксона: 42.0

p-значення: 0.0172

Оскільки p-значення менше за рівень значущості α = 0.05, є підстави відхилити нульову гіпотезу про відсутність змін у конверсії.

Отже, зміна рекламної кампанії призвела до статистично значущого підвищення показників конверсії. Це свідчить про позитивний вплив змін на поведінку користувачів.
Тест показав, що друга рекламна кампанія значно покращила конверсію.